### PR TITLE
[Eigensolvers] WIP AMatrix support

### PR DESCRIPTION
**Pull request by oberbichler**
- add support for AMatrix
- I need to redefine the Ublas vector in python for the tests (`KratosMultiphysics.Vector = AMatrix`)

### CODE REVIEW

**Comment by armingeiser** [comment]
The eigen solvers tests in the core are skipped, when i check why, the following exception is shown:

```
======================================================================
ERROR: test_eigen_eigensystem_solver (__main__.TestEigenSolvers)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_eigen_solvers.py", line 134, in test_eigen_eigensystem_solver
    raise e
  File "test_eigen_solvers.py", line 132, in test_eigen_eigensystem_solver
    import KratosMultiphysics.EigenSolversApplication
  File "/media/arming/Volume/04_Development/Kratos/KratosMultiphysics/EigenSolversApplication.py", line 12, in <module>
    from KratosEigenSolversApplication import *
ImportError: generic_type: type "UblasVector" is already registered!
```

**Comment by armingeiser** [comment]
The test is located in: `Kratos/kratos/tests/test_eigen_solvers.py`

**Comment by oberbichler** [comment]
This should only be a problem if you compile it without AMatrix. Then the core also exports the ublas vector to python -> multiple definition -> error.

Option A: the modification works only with AMatrix
Option B: it should work with ublas too

Of course I would prefer B because it will take some time until the switch to AMatrix is done.

Maybe we can use a macro to check whether AMatrix is used. But then we also need a way to check that on the python side (or just try…catch…).

**Comment by armingeiser** [comment]
Hm this seems to be a general problem of the porting to AMatrix. I don't think it makes sense to solve it for our application individually. The same issue will arise for the linear solvers in the core (at least for testing them).
So i think once it is clear how it is solved there, we can do the same thing in the EigenSolversApplication.

What i don't understand is that you exposed it as `UblasVector`. This term is never used anywhere else, so why does it say it is registered twice?

**Comment by oberbichler** [comment]
> So i think once it is clear how it is solved there, we can do the same thing in the EigenSolversApplication.

👍 

> What i don't understand is that you exposed it as UblasVector. This term is never used anywhere else, so why does it say it is registered twice?

Because internally python uses the type as an id. That means that you can not register the same class multiple times with different names.

File: applications/EigenSolversApplication/tests/test_eigen_direct_solver.py [hunk 1]
```diff
@@ -31,16 +31,16 @@ def _execute_eigen_direct_solver_test(self, class_name, solver_type):
 
         self.assertEqual(dimension, 900)
 
-        b_exp = KratosMultiphysics.Vector(dimension) # [1, 2, ..., dimension-1, dimension]
+        b_exp = EigenSolversApplication.UblasVector(dimension) # [1, 2, ..., dimension-1, dimension]
```
Comment (philbucher) [inline applications/EigenSolversApplication/tests/test_eigen_direct_solver.py:34]: Using the UblasSpace directly didn’t work?

Comment (oberbichler) [reply applications/EigenSolversApplication/tests/test_eigen_direct_solver.py:34]: I did not managed to test it. We decided to wait for a solution in the core.

File: applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h [hunk 2]
```diff
@@ -166,8 +166,8 @@ class EigenDirectSolver
      */
     void PerformSolutionStep(SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
     {
-        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > x(rX.data().begin(), rX.size());
-        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > b(rB.data().begin(), rB.size());
+        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > x(std::begin(rX.data()), rX.size());
+        Eigen::Map<Eigen::Matrix<DataType, Eigen::Dynamic, 1> > b(std::begin(rB.data()), rB.size());
```
Comment (pooyan-dadvand) [inline applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h:170]: nice solution! 👍

**Comment by jcotela** [comment]
@oberbichler I am assigning us two to the issue. This is only to indicate that I will be the member of the @KratosMultiphysics/technical-committee following this issue (we will try to do this for all issues on the Technical Committee project).

**Comment by philbucher** [comment]
just to mention that #3004 should solve the problem both for core and eigensolvers

**Review by pooyan-dadvand — COMMENTED** [review]
I understand that we should wait for #3004. Isn't it?

File: applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h (hunk 2 above)
Comment (oberbichler) [reply applications/EigenSolversApplication/custom_solvers/eigen_direct_solver.h:170]: thx!

**Comment by oberbichler** [comment]
@pooyan-dadvand Yes! And maybe also for #615.

**Pull request closed without merging**